src/MySqlConnector/MySql.Data.MySqlClient/MySqlConnection.cs

```diff
@@ -302,16 +302,10 @@ internal void Cancel(MySqlCommand command)
 			if (m_session.ServerVersion.Version < ServerVersions.SupportsProcedureCache)
 				return null;
 
-			if (m_cachedProcedures == null)
-				m_cachedProcedures = new Dictionary<string, CachedProcedure>();
-
 			var normalized = NormalizedSchema.MustNormalize(name, Database);
-			if (!m_cachedProcedures.TryGetValue(normalized.FullyQualified, out var cachedProcedure))
-			{
-				cachedProcedure = await CachedProcedure.FillAsync(ioBehavior, this, normalized.Schema, normalized.Component, cancellationToken).ConfigureAwait(false);
-				m_cachedProcedures[normalized.FullyQualified] = cachedProcedure;
-			}
-			return cachedProcedure;
+
+			return await m_cachedProcedures.GetOrAdd(normalized.FullyQualified,
+				(key) => CachedProcedure.FillAsync(ioBehavior, this, normalized.Schema, normalized.Component, cancellationToken));
```

Comment:
Capturing `this` will leak the connection and prevent the regular clean-up logic from detecting whether the user has abandoned the `MySqlConnection` object without closing it.

Reply:
If the `cancellationToken` is cancelled, a Canceled Task will be cached, preventing any future client from succeeding.

Reply:
Missing `.ConfigureAwait(false)`; this can lead to deadlock.